_source/_posts/2021-11-08-angular-v13.md

```diff
@@ -0,0 +1,183 @@
+
+Testing Angular apps is much improved with better cleanup of the DOM between tests defaulted in v13. Now, Angular does a better job of cleaning up the DOM between tests, and tests run faster and more performantly with better isolation between tests.
+
+The clean-up behavior is configurable when configuring or resetting the `TestBed`. Let's look at an example test where we greet the user after they've logged in.
+
+**Component code**
+
+{% raw %}
+```ts
+@Component({
+  selector: 'app-profile',
+  template: `
+    <ng-container *ngIf="name$ | async as name ">
+      <span>Welcome {{name}}!</span>
+    </ng-container>
+  `
+})
+
+export class ProfileComponent {
+  public name$: Observable<string> = this._authStateService.authState$.pipe(
+    filter((s: AuthState) => !!s && !!s.isAuthenticated),
+    map((s: AuthState) => s.idToken?.claims.name ?? 'stranger')
+  );
+
+  constructor(private _authStateService: OktaAuthStateService) { }
+}
+```
+{% endraw %}
+
+**Test code**
+
+```ts
+describe('ProfileComponent', () => {
+  let component: ProfileComponent;
+  let fixture: ComponentFixture<ProfileComponent>;
+
+  let authStateSpy = jasmine.createSpyObj<OktaAuthStateService>([], { authState$: of(authState) });
+
+  beforeEach(() => {
+    TestBed.configureTestingModule({
+      declarations: [ ProfileComponent ],
+      providers: [
+        { provide: OktaAuthStateService, useValue: authStateSpy }
+      ],
+      teardown: { destroyAfterEach: false } // flakey tests!
+    });
+  });
+
+  it('should show text that welcomes the user by name', () => {
+    fixture = TestBed.createComponent(ProfileComponent);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+      
+    const spanEl = fixture.debugElement.query(By.css('span'));
+    expect(spanEl).toBeTruthy();
+    expect(spanEl.nativeElement.innerHTML).toEqual('Welcome Test Name!');
+  });
+
+  it('should show text that welcomes a user with no name', () => {
+    (Object.getOwnPropertyDescriptor(authStateSpy, 'authState$')?.get as jasmine.Spy).and.returnValue(of({
+      isAuthenticated: true,
+      idToken: {...idToken, claims: {sub: 'sub'}}
+    }));
+      
+    fixture = TestBed.createComponent(ProfileComponent);
+    component = fixture.componentInstance;
+    fixture.detectChanges();
+
+    const spanEl = fixture.debugElement.query(By.css('span'));
+    expect(spanEl).toBeTruthy();
+    expect(spanEl.nativeElement.innerHTML).toEqual('Welcome stranger!');
+  });
+});
+```
+
+We can configure the teardown cleanup in this test suit by changing the value of `destroyAfterEach`. With this change enabled by default, you get the best option of writing new tests with better performance and disabling it for tests that aren't ready to consume the changes. Angular has our backs on non-breaking updates!
+
+## Enhanced developer tooling
+
+Creating Angular apps is now easier with enhancements to developer tooling, compilation time improvements, and updates to testbed environments.
```

Comment:
Should testbed be spelled as TestBed in this sentence?

Reply:
yes! something is not correct here. thanks for the catch.